finality-aleph/src/data_io.rs

```diff
@@ -203,138 +448,535 @@ where
+                .expect("client must answer a query")
+                .expect("block exists");
+            let parent_hash = *header.parent_hash();
+            self.on_block_imported(
+                BlockHashNum::<B>::new(client_info.best_hash, client_info.best_number),
+                parent_hash,
+            );
+        }
+        if self.highest_finalized.num < client_info.finalized_number {
+            self.on_block_finalized(BlockHashNum::<B>::new(
+                client_info.finalized_hash,
+                client_info.finalized_number,
+            ));
+        }
+
+        let proposals_with_timestamps: Vec<_> = self
+            .pending_proposals
             .iter()
-            .map(|(block, info)| (*block, info.first_occurence))
+            .map(|(proposal, info)| (proposal.clone(), info.first_occurence))
             .collect();
-        debug!(target: "afa", "Data Store maintanance. Awaiting {:?} blocks: {:?}", blocks_with_timestamps.len(), blocks_with_timestamps);
-        let finalized_number = self.client.info().finalized_number;
+        if !proposals_with_timestamps.is_empty() {
+            info!(target: "afa", "Data Store maintenance. Awaiting {:?} proposals.",proposals_with_timestamps.len());
+            let proposals_to_log = {
+                if proposals_with_timestamps.len() < 10 {
+                    proposals_with_timestamps.as_slice()
+                } else {
+                    &proposals_with_timestamps[..10]
+                }
+            };
+            info!(target: "afa", "Pending proposals (<=10): {:?}", proposals_to_log);
+        }
+
         let now = time::SystemTime::now();
-        for (block_data, first_occurence) in blocks_with_timestamps {
-            if matches!(
-                self.client.header(BlockId::Hash(block_data.hash)),
-                Ok(Some(_))
-            ) || finalized_number >= block_data.number
-            {
-                self.add_block(block_data);
-            } else if let Ok(time_waiting) = now.duration_since(first_occurence) {
-                if time_waiting >= self.config.request_block_after {
-                    debug!(target: "afa", "Requesting a stale block {:?} after it has been missing for {:?} secs.", block_data, time_waiting.as_secs());
-                    // Requesting a block multiple times is fine -- duplicate requests are not registered internally,
-                    // the old request is only updated. However, the substrate documentation suggests to not request
-                    // recent blocks this way (only stale forks), hence we wait at least 100 secs before attempting one.
-                    self.block_requester
-                        .request_stale_block(block_data.hash, block_data.number);
+        for (proposal, first_occurence) in proposals_with_timestamps {
+            if !self.bump_proposal(&proposal, true) {
+                if let Ok(time_waiting) = now.duration_since(first_occurence) {
```

Comment:
This should use `checked_duration_since` instead of `duration_since`.

Reply:
Huh, this is actually correct without changing anything, as it uses  `SystemTime` instead of `Instant` for which `duration_since` is checked by default.